backend/src/main/kotlin/org/loculus/backend/controller/FilesController.kt

```diff
@@ -0,0 +1,92 @@
+import io.swagger.v3.oas.annotations.security.SecurityRequirement
+import org.apache.http.HttpStatus
+import org.loculus.backend.api.AccessionVersion
+import org.loculus.backend.api.FileIdAndUrl
+import org.loculus.backend.auth.AuthenticatedUser
+import org.loculus.backend.auth.HiddenParam
+import org.loculus.backend.auth.User
+import org.loculus.backend.service.files.FilesDatabaseService
+import org.loculus.backend.service.files.S3Service
+import org.loculus.backend.service.groupmanagement.GroupManagementPreconditionValidator
+import org.loculus.backend.service.submission.AccessionPreconditionValidator
+import org.loculus.backend.service.submission.SubmissionDatabaseService
+import org.loculus.backend.utils.Accession
+import org.springframework.http.ResponseEntity
+import org.springframework.validation.annotation.Validated
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.RestController
+import java.net.URI
+
+@RestController
+@RequestMapping("/files")
+@Validated
+@SecurityRequirement(name = "bearerAuth")
+class FilesController(
+    private val filesDatabaseService: FilesDatabaseService,
+    private val s3Service: S3Service,
+    private val groupManagementPreconditionValidator: GroupManagementPreconditionValidator,
+    private val submissionDatabaseService: SubmissionDatabaseService,
+    private val accessionPreconditionValidator: AccessionPreconditionValidator,
+) {
+
+    @GetMapping("/get/{accession}/{version}/{fileCategory}/{fileName}")
+    fun getFileDownloadUrl(
+        @HiddenParam user: User,
+        @PathVariable accession: Accession,
+        @PathVariable version: Long,
+        @PathVariable fileCategory: String,
+        @PathVariable fileName: String,
+    ): ResponseEntity<Void> {
+        val accessionVersion = AccessionVersion(accession, version)
+        val fileId = submissionDatabaseService.getFileId(accessionVersion, fileCategory, fileName)
+        if (fileId == null) {
+            throw NotFoundException("File not found")
+        }
+        val isPublic = filesDatabaseService.isFilePublic(fileId)!!
+        if (!isPublic) {
+            if (user is AuthenticatedUser) {
+                accessionPreconditionValidator.validate {
+                    thatAccessionVersionExists(accessionVersion)
+                        .andThatUserIsAllowedToEditSequenceEntries(user)
+                }
+            } else {
+                throw ForbiddenException("Authentication needed: the file is not public")
+            }
+        }
+        val presignedUrl = s3Service.createUrlToReadPrivateFile(
+            fileId,
+            fileName,
+        )
+        return ResponseEntity.status(HttpStatus.SC_TEMPORARY_REDIRECT)
```

Comment:
Using a redirect means unless you inspect with curl or your browser dev tools or look at the download link in your browser one can't actually easily get that URL and share with others.

Every request to read a file, even a public file, has to go through backend

Reply:
> Every request to read a file, even a public file, has to go through backend

Yes - we discussed this in some detail on Slack: it means that each file will always be referenced with the same URL (which will be e.g. `http://instance.org/seq/PP_00140V0.1/raw_reads/forward.fastq.gz`). The fact that there is a single consistent URL that can persist over time, even if e.g. a new technology comes along that's not S3 or if we need to split storage across multiple regions is IMO a feature not a bug. It also means that files will download with the correct filename as we will generate a presigned URL that gives the file the right name in the attachment.